### PR TITLE
Unwrap year in comparison also for IN predicate

### PR DESCRIPTION
**Pull request by kabunchi**
## Release notes

(*) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Improve query performance for queries involving `year` function within `IN` predicate.
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java [hunk 1]
```diff
@@ -124,6 +129,36 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
             return unwrapYear(expression);
         }
 
+        @Override
+        public Expression rewriteInPredicate(InPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            Expression value = node.getValue();
+            Expression valueList = node.getValueList();
+            Expression inPredicate = treeRewriter.defaultRewrite((Expression) node, null);
```
Comment (kabunchi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java:137]: @ebyhr I am not sure I need this line. please advise.

**Comment by findepi** [comment]
> [force-pushed](https://github.com/trinodb/trino/compare/dc30ccfc052b5190bea0b4051f80384245ad25d8..600dba1bc24ab8e1d6dbbe24a5cffd8093462f55) the unwrap-in branch from [dc30ccf ](https://github.com/trinodb/trino/commit/dc30ccfc052b5190bea0b4051f80384245ad25d8)to [600dba1 ](https://github.com/trinodb/trino/commit/600dba1bc24ab8e1d6dbbe24a5cffd8093462f55)
[8 minutes ago](https://github.com/trinodb/trino/pull/18092#event-9722695187)


please avoid mixing changes with rebase. it's hard to notice what chaned.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java [hunk 2]
```diff
@@ -124,6 +129,36 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
             return unwrapYear(expression);
         }
 
+        @Override
+        public Expression rewriteInPredicate(InPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            Expression inPredicate = treeRewriter.defaultRewrite((Expression) node, null);
+            Expression value = node.getValue();
+            Expression valueList = node.getValueList();
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java:137]:
```suggestion
            Expression value = inPredicate.getValue();
            Expression valueList = inPredicate.getValueList();
```

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java [hunk 3]
```diff
@@ -124,6 +129,36 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
...
+                    call.getArguments().size() != 1 ||
+                    !(valueList instanceof InListExpression inListExpression)) {
+                return inPredicate;
+            }
+
+            // Convert each value to a comparison expression and try to unwrap it.
+            // unwrap the InPredicate only in case we manage to unwrap the entire value list
+            List<Expression> expressionsList = inListExpression.getValues();
+            ImmutableList.Builder<Expression> comparisonExpressionsListBuilder = ImmutableList.builder();
+            for (Expression currExpression : expressionsList) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java:150]: - avoid abbreviations ("curr")
- `comparisonExpressionsListBuilder` is a long name and the fact it's listbuilder is known from its type
- `expressionsList` can be inlined


```suggestion
            ImmutableList.Builder<Expression> comparisonExpressions = ImmutableList.builderWithExpectedSize(inListExpression.getValues().size());
            for (Expression rightExpression : inListExpression.getValues()) {
                ComparisonExpression comparisonExpression = new ComparisonExpression(EQUAL, value, rightExpression);
```

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java (hunk 2 above)
Comment (kabunchi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java:137]: done. I wasn't sure that the returned expression from `defaultRewriter` must be an `InPredicate` so I also added a validation.

File: core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapYearInComparison.java [hunk 4]
```diff
@@ -48,11 +48,13 @@ public void testEquals()
...
+        testUnwrap("date", "year(a) IN (1000, 1400, 1800)", "a BETWEEN DATE '1000-01-01' AND DATE '1000-12-31' OR a BETWEEN DATE '1400-01-01' AND DATE '1400-12-31' OR a BETWEEN DATE '1800-01-01' AND DATE '1800-12-31'");
 
         testUnwrap("timestamp", "year(a) = -0001", "a BETWEEN TIMESTAMP '-0001-01-01 00:00:00.000' AND TIMESTAMP '-0001-12-31 23:59:59.999'");
         testUnwrap("timestamp", "year(a) = 1960", "a BETWEEN TIMESTAMP '1960-01-01 00:00:00.000' AND TIMESTAMP '1960-12-31 23:59:59.999'");
         testUnwrap("timestamp", "year(a) = 2022", "a BETWEEN TIMESTAMP '2022-01-01 00:00:00.000' AND TIMESTAMP '2022-12-31 23:59:59.999'");
         testUnwrap("timestamp", "year(a) = 9999", "a BETWEEN TIMESTAMP '9999-01-01 00:00:00.000' AND TIMESTAMP '9999-12-31 23:59:59.999'");
+        testUnwrap("timestamp", "year(a) IN (1000, 1400, 1800)", "a BETWEEN TIMESTAMP '1000-01-01 00:00:00.000' AND TIMESTAMP '1000-12-31 23:59:59.999' OR a BETWEEN TIMESTAMP '1400-01-01 00:00:00.000' AND TIMESTAMP '1400-12-31 23:59:59.999' OR a BETWEEN TIMESTAMP '1800-01-01 00:00:00.000' AND TIMESTAMP '1800-12-31 23:59:59.999'");
```
Comment (ebyhr) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapYearInComparison.java:57]: This test method name is `testEquals`. Can we add `testIn`?

**Review by ebyhr — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapYearInComparison.java (hunk 4 above)
Comment (findepi) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapYearInComparison.java:57]: i don't see this addressed. BTW i pushed a small fixup, so please incorporate.

Comment (kabunchi) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapYearInComparison.java]: GitHub issue... now all comments should be incorporated.

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
@kabunchi thanks, will merge when the build passes
can you please follow-up with https://github.com/trinodb/trino/pull/18092/files#r1252095563 ?

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java [hunk 5]
```diff
@@ -124,6 +129,36 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
             return unwrapYear(expression);
         }
 
+        @Override
```
[the remark this answers is not in the document]
Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapYearInComparison.java:131]: Follow-up https://github.com/trinodb/trino/pull/18138

**Pull request merged**